### PR TITLE
Demand components

### LINKED ISSUE #33: [FEATURE] Add Truck
### User Story
*As a ...insert type of user... I'd like to ...insert desired feature or behavior...*

### Project
*Is there a funder or project associated with this feature?*

### Who should be involved?
Implementer: 
Commenters: 
Users: 
Reviewers: 

### Risk
*Will this potentially break anything?*

#### Tests
*What are relevant tests or what tests need to be created in order to determine that this issue is complete?*

### LINKED ISSUE #34: [FEATURE] Add internal / external trips.
### User Story
*As a model user,  I'd like to run the internal/external trips through `tm2py`.*

### Checklist

- [ ] User story satisfied
- [ ] Doc strings
- [ ] General documentation
- [ ] Tests
- [ ] Passing tests

### Level of Effort
Medium

### Resolution Ideas
Add as a series of components in internal external module

### Who should be involved?
Commenters: 
Users: 
Reviewers: 

### Risk
Low

#### Tests
*What are relevant tests or what tests need to be created in order to determine that this issue is complete?*

- [ ] Add relevant component tests

### LINKED ISSUE #59: [FEATURE] Add air passenger trip model
### User Story
*As a model user,  I'd like to run the air passenger trips through `tm2py`.*

### Checklist

- [ ] User story satisfied
- [ ] Doc strings
- [ ] General documentation
- [ ] Tests
- [ ] Passing tests

### Level of Effort
Medium

### Resolution Ideas
Add as a series of components in internal external module

### Who should be involved?
Commenters: 
Users: 
Reviewers: 

### Risk
Low

#### Tests
*What are relevant tests or what tests need to be created in order to determine that this issue is complete?*

- [ ] Add relevant component tests

### PR DESCRIPTION
**Pull request by inrokevin**
## Adding main non-resident demand components
Closes #59 - Air passenger trip model
Closes #34 - Internal/External trip model
Closes #33 - Truck model

## Testing plan
 - [X] Code for this PR is covered in tests, basic model run test added which executes these steps together with highway (transit to be added when branches are merged together)
 - [X] Code passes all existing tests
 - [X] Code linted

### CODE REVIEW

File: tm2py/emme/matrix.py [hunk 1]
```diff
@@ -262,3 +262,91 @@ def read_hdf5(self, path: str) -> NumpyArray:
...
+    from OMXManager
+
+    Args:
+        value_of_time: value of time to use in the utility expression
+        coeff_time: coefficient of time value used in the utility expression
+        operating_cost_per_mile: operating cost value in cents per mile
+            for converting distance to cost
+    """
+
+    def __init__(
```
Comment (e-lo) [inline tm2py/emme/matrix.py:281]: docstrings for __init__ args go here. 
docstrings for instance attributes go in class-level docstrings

File: tm2py/emme/matrix.py [hunk 2]
```diff
@@ -262,3 +262,91 @@ def read_hdf5(self, path: str) -> NumpyArray:
...
+        Loads the referenced skim matrices and calculates the result as:
+        exp(coeff_time * time + coeff_cost * (op_cost * dist + cost)))
+
+        coeff_cost = coeff_time / vot * 0.6
+
+        Args:
+            time_name: Name of the time skim matrix in the OMX file
+            dist_name: Name of the distance skim matrix in the OMX file
+            toll_names: List of names of the the toll skim matrix in the OMX file
+            toll_factor: Optional factor to apply to the tolls
```
Comment (e-lo) [inline tm2py/emme/matrix.py:316]: more description please - this is multiplied by the tolls? what is an example use case?

File: tm2py/emme/matrix.py [hunk 3]
```diff
@@ -262,3 +262,91 @@ def read_hdf5(self, path: str) -> NumpyArray:
...
+
+        Returns:
+            A numpy array with the calculated exp(util) result.
```
Comment (e-lo) [inline tm2py/emme/matrix.py:319]: add dimensions/units for the numpy array

File: tm2py/emme/matrix.py [hunk 4]
```diff
@@ -262,3 +262,91 @@ def read_hdf5(self, path: str) -> NumpyArray:
...
+        dist = self._omx_manager.read(dist_name)
+        toll = self._omx_manager.read(toll_names[0])
+        for name in toll_names[1:]:
+            toll += self._omx_manager.read(name)
+        if toll_factor != 1:
+            toll = toll * toll_factor
+        e_util = exp(k_ivtt * time + k_cost * (op_cost * dist + toll))
+        return e_util
+
+    def mask_non_available(
```
Comment (e-lo) [inline tm2py/emme/matrix.py:335]: This seems like it should be a "general step" for every mode - why is it special for tolls?
2. Shouldn't this be reading in and returning a masked skim?
3. Where are probabilities for tolls zero-ed out? It seems like it should be in same step.

File: tm2py/emme/matrix.py [hunk 5]
```diff
@@ -262,3 +262,91 @@ def read_hdf5(self, path: str) -> NumpyArray:
             Numpy array from OMX file
         """
         return self._omx_file.get_node(path).read()
+
+
+class TollChoiceCalculator:
```
Comment (e-lo) [inline tm2py/emme/matrix.py:267]: why is this in emme/matrix? Does it use Emme?

Comment (e-lo) [reply tm2py/emme/matrix.py:267]: i.e. why isn't this a model component which could then reference some general matrix/choice model language/tools

File: tm2py/config.py [hunk 6]
```diff
@@ -129,15 +140,49 @@ class AirPassengerDemandAggregationConfig(ConfigItem):
...
+    zone_index: int
+    factor: float
+
+
+@dataclass(frozen=True)
+class TimeOfDaySplitConfig(ConfigItem):
+    """Time of day demand split for productions and attractions"""
+
+    time_period: str
+    production: float
```
Comment (e-lo) [inline tm2py/config.py:182]: add requirements (i.e. between 0-1, sum of both are 1)

File: tm2py/config.py [hunk 7]
```diff
@@ -129,15 +140,49 @@ class AirPassengerDemandAggregationConfig(ConfigItem):
...
+
+    zone_index: int
+    factor: float
+
+
+@dataclass(frozen=True)
+class TimeOfDaySplitConfig(ConfigItem):
+    """Time of day demand split for productions and attractions"""
+
+    time_period: str
```
Comment (e-lo) [inline tm2py/config.py:181]: add requirements

File: tm2py/config.py [hunk 8]
```diff
@@ -129,15 +140,49 @@ class AirPassengerDemandAggregationConfig(ConfigItem):
...
     reference_end_year: str
+    airport_names: Tuple[str, ...]
     demand_aggregation: Tuple[AirPassengerDemandAggregationConfig, ...]
 
 
+@dataclass(frozen=True)
+class GateFactorConfig(ConfigItem):
+    """Mapping of gateway (zone ordering ID) to factor value"""
+
+    zone_index: int
```
Comment (e-lo) [inline tm2py/config.py:173]: add requirement

File: tm2py/config.py [hunk 9]
```diff
@@ -129,15 +140,49 @@ class AirPassengerDemandAggregationConfig(ConfigItem):
...
+    factor: float
```
Comment (e-lo) [inline tm2py/config.py:174]: add requirement.

File: tm2py/config.py [hunk 10]
```diff
@@ -129,15 +140,49 @@ class AirPassengerDemandAggregationConfig(ConfigItem):
...
     input_demand_folder: str
     reference_start_year: str
     reference_end_year: str
+    airport_names: Tuple[str, ...]
     demand_aggregation: Tuple[AirPassengerDemandAggregationConfig, ...]
 
 
+@dataclass(frozen=True)
+class GateFactorConfig(ConfigItem):
+    """Mapping of gateway (zone ordering ID) to factor value"""
```
Comment (e-lo) [inline tm2py/config.py:171]: what the heck is a "factor value"? :-)

File: tm2py/config.py [hunk 11]
```diff
@@ -129,15 +140,49 @@ class AirPassengerDemandAggregationConfig(ConfigItem):
...
+        in the file name template {year}_{direction}{airport}.csv
+    airport_names: list of one or more airport names / codes as used in
+        the input file names
+    demand_aggregation: specification of aggregation of by-access mode
+        demand to highway class demand
+    """
 
     highway_demand_file: str
     input_demand_folder: str
     reference_start_year: str
```
Comment (e-lo) [inline tm2py/config.py:163]: add requirement
(and why a string not an int?)

File: tm2py/config.py [hunk 12]
```diff
@@ -129,15 +140,49 @@ class AirPassengerDemandAggregationConfig(ConfigItem):
...
     reference_end_year: str
```
Comment (e-lo) [inline tm2py/config.py:164]: ditto

File: tm2py/config.py [hunk 13]
```diff
@@ -120,7 +121,17 @@ class HouseholdConfig(ConfigItem):
...
+        result_class_name: name used in the output OMX matrix names, note
+            that this should match the expected naming convention in the
+            HighwayClassDemandConfig name(s)
+        src_group_name: name used for the class group in the input columns
+            for the trip tables,
+        access_modes: list of names used for the access modes in the input
+            columns for the trip tables
+    """
 
     result_class_name: str
```
Comment (e-lo) [inline tm2py/config.py:136]: add requirement, should be one of....

File: tm2py/config.py [hunk 14]
```diff
@@ -120,7 +121,17 @@ class HouseholdConfig(ConfigItem):
...
     src_group_name: str
```
Comment (e-lo) [inline tm2py/config.py:137]: add requirement, should be one of...

File: tm2py/config.py [hunk 15]
```diff
@@ -120,7 +121,17 @@ class HouseholdConfig(ConfigItem):
...
 @dataclass(frozen=True)
 class AirPassengerDemandAggregationConfig(ConfigItem):
-    """Air passenger demand aggregation input parameters"""
+    """Air passenger demand aggregation input parameters.
+
+    Properties:
+        result_class_name: name used in the output OMX matrix names, note
+            that this should match the expected naming convention in the
+            HighwayClassDemandConfig name(s)
+        src_group_name: name used for the class group in the input columns
```
Comment (e-lo) [inline tm2py/config.py:130]: what is the meaning of "class group"? (example would help)

File: tm2py/config.py [hunk 16]
```diff
@@ -120,7 +121,17 @@ class HouseholdConfig(ConfigItem):
 
 @dataclass(frozen=True)
 class AirPassengerDemandAggregationConfig(ConfigItem):
-    """Air passenger demand aggregation input parameters"""
+    """Air passenger demand aggregation input parameters.
+
+    Properties:
+        result_class_name: name used in the output OMX matrix names, note
+            that this should match the expected naming convention in the
+            HighwayClassDemandConfig name(s)
```
Comment (e-lo) [inline tm2py/config.py:129]: Add example

File: tm2py/config.py [hunk 17]
```diff
@@ -120,7 +121,17 @@ class HouseholdConfig(ConfigItem):
...
+        src_group_name: name used for the class group in the input columns
+            for the trip tables,
+        access_modes: list of names used for the access modes in the input
```
Comment (e-lo) [inline tm2py/config.py:132]: don't see this there?

File: tests/test_union_city.py [hunk 18]
```diff
@@ -53,10 +55,12 @@ def test_example_download():
...
     for file_name in files_to_check:
         assert os.path.exists(
-            os.path.join(example_dir, name, file_name)
+            os.path.join(_UNION_CITY_DIR, file_name)
         ), f"get_example failed, missing {file_name}"
     # check zip file was removed
-    assert not (os.path.exists(os.path.join(example_dir, name, "test_data.zip")))
+    assert not (
+        os.path.exists(os.path.join(_UNION_CITY_DIR, "test_data.zip"))
+    )
```
Comment (e-lo) [inline tests/test_union_city.py:65]: There aren't any tests for any of these submodules.

File: tm2py/components/demand/air_passenger.py [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+
+import numpy as np
+import openmatrix as _omx
+import pandas as pd
+
+from tm2py.components.component import Component
+from tm2py.logger import LogStartEnd
+
+if TYPE_CHECKING:
+    from tm2py.controller import RunController
```
Comment (e-lo) [inline tm2py/components/demand/air_passenger.py:16]: oh fancy! i like this !(assuming it works)

File: tm2py/components/demand/air_passenger.py [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+        super().__init__(controller)
+        self._periods = [p.upper() for p in self.time_period_names()]
+        self._start_year = None
+        self._end_year = None
+        self._mode_groups = {}
+        self._out_names = {}
+
+    @LogStartEnd()
+    def run(self):
+        """Build the airport trip matrices"""
```
Comment (e-lo) [inline tm2py/components/demand/air_passenger.py:120]: add docstring about steps it is doing here

File: tm2py/components/demand/air_passenger.py [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+        for year in [self._start_year, self._end_year]:
+            for period in self._periods:
+                for group, access_modes in self._mode_groups.items():
+                    data = input_demand[
+                        [f"{period}_{access}_{group}_{year}" for access in access_modes]
+                    ]
+                    aggr_demand[f"{period}_{group}_{year}"] = data.sum(axis=1)
+        return aggr_demand
+
+    def _interpolate(self, aggr_demand: pd.DataFrame) -> pd.DataFrame:
```
Comment (e-lo) [inline tm2py/components/demand/air_passenger.py:175]: This should be a common element shared for components.

Comment (e-lo) [reply tm2py/components/demand/air_passenger.py:175]: [ and have a unit-test ]

File: tm2py/components/demand/air_passenger.py [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+        """Interpolate for the model year assuming linear growth between the reference years."""
+        start_year = int(self._start_year)
```
Comment (e-lo) [inline tm2py/components/demand/air_passenger.py:177]: these should be args

File: tm2py/components/demand/air_passenger.py [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+        demand = pd.DataFrame()
+        for period in self._periods:
+            for group in self._mode_groups:
+                name = f"{period}_{group}"
+                demand[name] = (1 - scale) * aggr_demand[
+                    f"{name}_{start_year}"
+                ] + scale * aggr_demand[f"{name}_{end_year}"]
+        return demand
+
+    def _export_result(self, demand: pd.DataFrame):
```
Comment (e-lo) [inline tm2py/components/demand/air_passenger.py:205]: probably also a common element

File: tm2py/components/demand/air_passenger.py [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+        """Loads demand from the CSV files into pandas dataframe."""
+
+        def rename_columns(name):
+            if name in ["ORIG", "DEST"]:
+                return name
+            return f"{name}_{year}"
+
+        input_data_folder = self.config.air_passenger.input_demand_folder
+        input_demand = []
+        for year in [self._start_year, self._end_year]:
```
Comment (e-lo) [inline tm2py/components/demand/air_passenger.py:144]: itertools.product()?

File: tm2py/components/demand/air_passenger.py [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+        for group in self.config.air_passenger.demand_aggregation:
+            self._mode_groups[group["src_group_name"]] = group["access_modes"]
+            self._out_names[group["src_group_name"]] = group["result_class_name"]
+
+        input_demand = self._load_demand()
+        aggr_demand = self._aggregate_demand(input_demand)
+        demand = self._interpolate(aggr_demand)
+        self._export_result(demand)
+
+    def _load_demand(self) -> pd.DataFrame:
```
Comment (e-lo) [inline tm2py/components/demand/air_passenger.py:134]: is there anything special here? if so, suggest _load_air_demand() or similar.

File: tm2py/components/demand/air_passenger.py [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+                return name.replace(f"_{year}", "")
+
+            demand.rename(columns=rename_columns, inplace=True)
+            return demand
+
+        # In the cube .job script, the formula is:
+        # token_scale = (%MODEL_YEAR% - 2007)/(2035 - %MODEL_YEAR%)
+        # and it should be: token_scale = (%MODEL_YEAR% - 2007)/(2035 - 2007)
+        # scale = float((int(year) - 2007)) / (2035 - int(year))
+        scale = float(int(year) - start_year) / (end_year - start_year)
```
Comment (e-lo) [inline tm2py/components/demand/air_passenger.py:195]: scale demand seems like it should be a common function...

File: tm2py/components/demand/commercial.py [hunk 27]
```diff
@@ -0,0 +1,611 @@
...
+from tm2py.tools import parse_num_processors
+
+if TYPE_CHECKING:
+    from tm2py.controller import RunController
+
+NumpyArray = np.array
+
+
+# employment category mappings, grouping into larger categories
+_land_use_aggregation = {
```
Comment (e-lo) [inline tm2py/components/demand/commercial.py:22]: a param?

File: tm2py/components/demand/commercial.py [hunk 28]
```diff
@@ -0,0 +1,611 @@
...
+        MWTEMPN, OTHEMPN, TOTEMP, TOTHH
+
+        Returned columns are: vsmtrk_prod, vsmtrk_attr, smltrk_prod,
+        smltrk_attr, medtrk_prod, medtrk_attr, lrgtrk_prod, lrgtrk_attr
+        """
+
+        link_trips = pd.DataFrame()
+        # linked trips (non-garage-based) - productions
+        # (very small updated with NAICS coefficients)
+        link_trips["vsmtrk_prod"] = (
```
Comment (e-lo) [inline tm2py/components/demand/commercial.py:282]: param?

File: tm2py/components/demand/commercial.py [hunk 29]
```diff
@@ -0,0 +1,611 @@
...
+        link_trips["medtrk_prod"] = (0.0039 * landuse.TOTEMP).round()
+        link_trips["lrgtrk_prod"] = (0.0073 * landuse.TOTEMP).round()
+        # linked trips (non-garage-based) - attractions (equal productions)
+        link_trips["vsmtrk_attr"] = link_trips["vsmtrk_prod"]
+        link_trips["smltrk_attr"] = link_trips["smltrk_prod"]
+        link_trips["medtrk_attr"] = link_trips["medtrk_prod"]
+        link_trips["lrgtrk_attr"] = link_trips["lrgtrk_prod"]
+        garage_trips = pd.DataFrame()
+        # garage-based - productions (updated NAICS coefficients)
+        garage_trips["smltrk_prod"] = (
```
Comment (e-lo) [inline tm2py/components/demand/commercial.py:304]: param?

File: tm2py/components/demand/commercial.py [hunk 30]
```diff
@@ -0,0 +1,611 @@
...
+                f"{name}, prod sum: {prod_attr_matrix.sum()}, "
+                f"daily sum: {daily_demand[name].sum()}",
+                level="DEBUG",
+            )
+        return daily_demand
+
+    def _get_blended_time(self, name: str) -> NumpyArray:
+        """Blend AM and MD truck times for distribution calculations.
+
+        Uses 1/3 * AM + 2/3 * MD, sources skims from generated OMX files
```
Comment (e-lo) [inline tm2py/components/demand/commercial.py:418]: make the blending factors a param?

File: tm2py/components/demand/commercial.py [hunk 31]
```diff
@@ -0,0 +1,611 @@
...
+            )
+            self.logger.log(
+                f"{name}, prod sum: {prod_attr_matrix.sum()}, "
+                f"daily sum: {daily_demand[name].sum()}",
+                level="DEBUG",
+            )
+        return daily_demand
+
+    def _get_blended_time(self, name: str) -> NumpyArray:
+        """Blend AM and MD truck times for distribution calculations.
```
Comment (e-lo) [inline tm2py/components/demand/commercial.py:416]: make time periods a param?

File: tm2py/components/demand/commercial.py [hunk 32]
```diff
@@ -0,0 +1,611 @@
...
+        #           are in the highway_skims_{period}.omx files in Emme version
+        #           with updated matrix names, {period}_trk_time, {period}_lrgtrk_time.
+        #           Also, there will no longer be separate very small, small and medium
+        #           truck times, as they are assigned together as the same class.
+        #           There is only the trk_time.
+        blend_time = 1 / 3.0 * am_time + 2 / 3.0 * md_time
+        return blend_time
+
+    @LogStartEnd(level="DEBUG")
+    def _time_of_day(
```
Comment (e-lo) [inline tm2py/components/demand/commercial.py:437]: this function name should  be an action, not a noun - ex: `_split_by_time_period

File: tm2py/components/demand/commercial.py [hunk 33]
```diff
@@ -0,0 +1,611 @@
...
+        self, daily_demand: Dict[str, Dict[str, NumpyArray]]
```
Comment (e-lo) [inline tm2py/components/demand/commercial.py:438]: demand, factors should be params?

File: tm2py/components/demand/commercial.py (hunk 32 above)
Comment (e-lo) [inline tm2py/components/demand/commercial.py:437]: Does this happen other places? seems like a common method which could be specified in one place.

File: tm2py/components/demand/commercial.py [hunk 34]
```diff
@@ -0,0 +1,611 @@
...
+        for period in self.time_period_names():
+            factor_map = _time_of_day_split[period]
+            demand = {}
+            for name, factor in factor_map.items():
+                demand[name] = np.around(factor * daily_demand[name], decimals=2)
+            period_demand[period] = demand
+        return period_demand
+
+    @LogStartEnd(level="DEBUG")
+    def _toll_choice(
```
Comment (e-lo) [inline tm2py/components/demand/commercial.py:460]: why is this here but here is another one in emme/matrix?

If this is common function, should pull out.

File: tm2py/components/demand/household.py [hunk 35]
```diff
@@ -0,0 +1,56 @@
...
+        self._start_household_manager()
+        self._start_matrix_manager()
+        self._run_resident_model()
+        self._stop_java()
+
+    @staticmethod
+    def _start_household_manager():
+        commands = [
+            "CALL CTRAMP\\runtime\\CTRampEnv.bat",
+            "set PATH=%CD%\\CTRAMP\runtime;C:\\Windows\\System32;%JAVA_PATH%\bin;"
```
Comment (e-lo) [inline tm2py/components/demand/household.py:25]: make paths safe

File: tm2py/components/demand/household.py [hunk 36]
```diff
@@ -0,0 +1,56 @@
...
+
+import shutil as _shutil
+
+from tm2py.components.component import Component
+from tm2py.tools import run_process
+from tm2py.logger import LogStartEnd
+
+
+class HouseholdModel(Component):
+    """Run household resident model"""
```
Comment (e-lo) [inline tm2py/components/demand/household.py:12]: document more

File: tm2py/components/demand/internal_external.py [hunk 37]
```diff
@@ -0,0 +1,292 @@
...
+                  Value-toll bridge toll is: {period}_{class}toll_bridgetoll_{class},
+                  Value-toll value toll is: {period}_{class}toll_valuetoll_{class},
+
+     Output: Five, six-table trip matrices, one for each time period.  Two tables for each vehicle
+             class representing value-toll paying path trips and non-value-toll paying path trips.
+    """
+
+    @LogStartEnd()
+    def run(self):
+        """docstring for component run"""
```
Comment (e-lo) [inline tm2py/components/demand/internal_external.py:127]: add docstring

File: tm2py/components/demand/internal_external.py [hunk 38]
```diff
@@ -0,0 +1,292 @@
...
+        omx_file = _omx.open_file(file_path)
+        demand = {
+            "da": omx_file["IX_Daily_DA"].read(),
+            "sr2": omx_file["IX_Daily_SR2"].read(),
+            "sr3": omx_file["IX_Daily_SR3"].read(),
+        }
+        omx_file.close()
+        return demand
+
+    def _growth_forecast(self, demand: Dict[str, NumpyArray]) -> Dict[str, NumpyArray]:
```
Comment (e-lo) [inline tm2py/components/demand/internal_external.py:150]: seems like a good common function?

File: tm2py/components/demand/internal_external.py [hunk 39]
```diff
@@ -0,0 +1,292 @@
...
+        for entry in self.config.internal_external.time_of_day_split:
+            class_demand = {}
+            for class_name, demand in daily_prod_attract.items():
+                prod, attract = 0.5 * entry.production, 0.5 * entry.attraction
+                class_demand[class_name] = prod * demand + attract * demand.T
+            period_demand[entry.time_period] = class_demand
+        return period_demand
+
+    @LogStartEnd()
+    def _toll_choice(
```
Comment (e-lo) [inline tm2py/components/demand/internal_external.py:215]: ditto from other

**Review by e-lo — CHANGES_REQUESTED** [review]
Min needs to be addressed: add tests (there are none!) and missing docstrings
Preferable (although could be a refactor later...@lmz..) is to 
a. pull out common functions to be referenced by many processes
b. parameterize in parameters, not in magic numbers

File: tm2py/components/demand/air_passenger.py (hunk 20 above)
Comment (e-lo) [reply tm2py/components/demand/air_passenger.py:120]: fixed

**Comment by lmz** [comment]
Is this pull request still in progress or can it be merged?

**Comment by lmz** [comment]
Hello - I'm taking a deeper look at this code and testing it and I do have some thoughts/questions

1. I'd like to make logging not require a context; I find it cumbersome and adds indent clutter, which makes logging feel like a big deal
2. I don't think it's necessary to include timestamp as an option in logging. Let's just include timestamps.
3. I don't understand the `log_on_error_file_path`?  What is this file?
4. Likewise, what's the `LogCache` for?

I'm making the changes for 1 and 2 but would appreciate clarification on 3 and 4.  Thank you!

**Comment by lmz** [comment]
Hi @inrokevin @e-lo (@AntoineGrapperon in case @inrokevin is not available) -
Just pinging on my two questions above: What is LogCache and log_on_error_file_path for?
I don't want the logging system to be over-complicated, and I think two log files seem like they should be sufficient; it seems like error messages should be in both of the other two logs so why the need for a 3rd, and why the need for a LogCache system?

**Comment by e-lo** [comment]
As I think I've said before, I agree with @lmz that we shouldn't have to use a context - see #83 for the defined problem.

Since I am not really familiar with how @inrokevin implemented the context and what the referenced `LogCache` and `log_on_error_file_path` are for because I was hoping we would fix it wholesale. 

Do you think we need to:
(1) *fix* the logging with context in this PR? or 
(2) can we save this as an issue for now and not have it blocking the merge?

My preference in order to get things done now is 2...but you're the boss!

**Comment by lmz** [comment]
Got it. 

I opened #87 regarding the `LogCache` and `log_on_error_file_path`.  Although I just noted the description [in the doc here]( https://bayareametro.github.io/tm2py/demand_components/contributing/development/#viewing-logging) which refers to the `log_on_error_file_path` as pointing to the "Catch-all Log".  So from this standpoint, it defers from the detailed log or debug log in that it has trace as well?  But it also doesn't seem to always trigger?  In any case, I'm ok with punting on issue #87 for now.

 #83 bothers me too much so I checked in a fix (not 100% sure it's complete given my confusion around #87).

**Comment by i-am-sijia** [comment]
Passing tests:

- tests/test_air_access.py
- tests/test_trucks.py
- tests/test_externals.py

Example union city data: [UnionCity.zip](https://github.com/BayAreaMetro/tm2py/blob/demand_components/examples/UnionCity.zip)

**Comment by e-lo** [comment]
@i-am-sijia it looks like somewhere the github runner is assigning the wrong value to `controller.run_dir`.  Locally I get: `Users/elizabeth/Documents/urbanlabs/working/tm2py/examples/UnionCity`
GH Actions seems to be getting a double `tm2py` 

Suggest troublshooting the below conftest.py fixture to see if it needs to be more resilient:

```python
@pytest.fixture(scope="session")
def root_dir():
    """Root tm2py directory."""
    return Path(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
```

**Comment by i-am-sijia** [comment]
@e-lo I don't think the path in GH Actions is wrong. Looking at the log, it was able to find other files with the additional `tm2py` in path:

image

I think the problem is that this log file is not created:
image

**Comment by i-am-sijia** [comment]
@e-lo I think I found the reason... `'/home/runner/work/tm2py/tm2py/examples/UnionCity/'` folder does not exist because some test does not run the get example data step.

**Comment by i-am-sijia** [comment]
Passed all local tests. GH Actions are also passed. Ready to merge into Develop.
